pkix/key.go

```diff
@@ -61,42 +92,63 @@ func NewKeyFromPrivateKeyPEM(data []byte) (*Key, error) {
 	if pemBlock == nil {
 		return nil, errors.New("cannot find the next PEM formatted block")
 	}
-	if pemBlock.Type != rsaPrivateKeyPEMBlockType || len(pemBlock.Headers) != 0 {
-		return nil, errors.New("unmatched type or headers")
-	}
-
-	priv, err := x509.ParsePKCS1PrivateKey(pemBlock.Bytes)
-	if err != nil {
-		return nil, err
+	var signer crypto.Signer
+	switch pemBlock.Type {
+	case rsaPrivateKeyPEMBlockType:
+		priv, err := x509.ParsePKCS1PrivateKey(pemBlock.Bytes)
+		if err != nil {
+			return nil, err
+		}
+		signer = priv
+	case pkcs8PrivateKeyPEMBlockType:
+		priv, err := x509.ParsePKCS8PrivateKey(pemBlock.Bytes)
+		if err != nil {
+			return nil, err
+		}
+		signer = priv.(crypto.Signer)
+	default:
+		return nil, fmt.Errorf("unknown PEM block type %q", pemBlock.Type)
 	}
-
-	return NewKey(&priv.PublicKey, priv), nil
+	return NewKeyFromSigner(signer), nil
 }
 
-// NewKeyFromEncryptedPrivateKeyPEM inits Key from encrypted PEM-format rsa private key bytes
+// NewKeyFromEncryptedPrivateKeyPEM inits Key from encrypted PEM-format private key bytes
 func NewKeyFromEncryptedPrivateKeyPEM(data []byte, password []byte) (*Key, error) {
 	pemBlock, _ := pem.Decode(data)
 	if pemBlock == nil {
 		return nil, errors.New("cannot find the next PEM formatted block")
 	}
-	if pemBlock.Type != rsaPrivateKeyPEMBlockType {
-		return nil, errors.New("unmatched type or headers")
-	}
-
-	b, err := x509.DecryptPEMBlock(pemBlock, password)
-	if err != nil {
-		return nil, err
-	}
-
-	priv, err := x509.ParsePKCS1PrivateKey(b)
-	if err != nil {
-		return nil, err
+	var signer crypto.Signer
+	switch pemBlock.Type {
+	case rsaPrivateKeyPEMBlockType:
+		b, err := x509.DecryptPEMBlock(pemBlock, password)
+		if err != nil {
+			return nil, err
+		}
+		priv, err := x509.ParsePKCS1PrivateKey(b)
+		if err != nil {
+			return nil, err
+		}
+		signer = priv
+	case encryptedPKCS8PrivateKeyPEMBLockType:
+		b, err := pemutil.DecryptPKCS8PrivateKey(pemBlock.Bytes, password)
```

Comment:
is this something we need this 3p library for? (so nothing in stdlib or `x/crypto` that we can use?)

Reply:
Unfortunately not, see this discussion: https://github.com/golang/go/issues/8860#issuecomment-66099274

That thread is where I found this package, which looks like it may get considered for inclusion into `x/crypto`.